### ISSUE TITLE
OMEMO implementation error, 

### ISSUE THREAD

**Issue by the-eater** [dino/dino#303]
Currently while implementing OMEMO support for stanza.io, I hit an issue where I wasn't able to decrypt the OMEMO messages sent from dino.

I calculate the auth tag length of the AES-GCM ciphertext, by grabbing the length of the whisper message plaintext and subtract that by the key size, defined by XEP-0384, 128 bits. but on dino this constantly resulted in 0. with the only logical conclusion being that the auth tag is not appended to the WhipserMessage, even while XEP-0384 states that it is to be concatenated with the key for the ciphertext.

After hardcoding the tagLength to 128 bits, decryption worked. confirming the fact that the auth tag is not removed from the ciphertext and appended to the WhipserMessage.

**TL;DR;**

in an OMEMO message the auth tag is not removed from the cipher text and appended to the WhisperMessage, resulting in a faulty implementation of XEP-0384

**Comment by mar-v-in** [comment]
Yes, dino currently sends omemo messages in ProtoXEP legacy format. It supports receiving messages in legacy and XEP-384 format and will eventually switch over to the new format. If you don't want to support the legacy format, you can test against Conversations which uses the XEP format for sending messages since mid-2017.

**Comment by the-eater** [comment]
Since XEP-384 says the tag SHOULD be a minimum of 128 bits, I currently run `Math.max(128, tagLength)`, on it. which also solves the problem.

is there any documentation of the ProtoXEP format?

**Comment by mar-v-in** [comment]
I think [this](https://conversations.im/xeps/multi-end.html) is the original ProtoXEP version, it is not very clear about what to do with the auth tag, but all implementations a year ago basically had the behavior we implemented in Dino. You can see what Conversations did before [here](https://github.com/siacs/Conversations/blob/master/src/main/java/eu/siacs/conversations/crypto/axolotl/XmppAxolotlMessage.java#L174). The legacy format is intended to fade out, so you may as well completely ignore it - we should move to sending using the new format as well (because messages are not authenticated if we don't :wink:)

**Comment by the-eater** [comment]
The funny thing about AES-GCM is that once you finalize the stream, it will put the authentication tag at the end. so currently it's just being authenticated fine :) the only problem is that there is no way to know how long it is. while that is possible with XEP-384

**Comment by mar-v-in** [comment]
@the-eater maybe I am wrong, but the authentication tag does not properly authenticate if it is only appended to the ciphertext and multiple users do have the message key (which is the case with OMEMO in multi user chats).

If someone with the message key does MITM, they can encrypt a new message and replace the original cipertext+authtag with a new one. However if the authtag is attached to the key (like in XEP), any change in the ciphertext can be detected and the attacker will not be able to change the authtag because the key+authtag is e2e-encrypted.

**Comment by mar-v-in** [comment]
This was a duplicate of #514 and thus fixed in cc7b0aa7bd5b6599159f654fdd8a2fd111e16a3e.

**Comment by the-eater** [comment]
I think it was the other way around, but thanks
